src/AspNet.Security.OAuth.Alipay/AlipayAuthenticationHandler.cs

```diff
@@ -0,0 +1,169 @@
+            }
+            string address = QueryHelpers.AddQueryString(Options.UserInformationEndpoint, signparames);
+
+            using var response = await Backchannel.GetAsync(address);
+            if (!response.IsSuccessStatusCode)
+            {
+                Logger.LogError("An error occurred while retrieving the user profile: the remote server " +
+                                "returned a {Status} response with the following payload: {Headers} {Body}.",
+                                /* Status: */ response.StatusCode,
+                                /* Headers: */ response.Headers.ToString(),
+                                /* Body: */ await response.Content.ReadAsStringAsync());
+
+                throw new HttpRequestException("An error occurred while retrieving user information.");
+            }
+            using var payload = JsonDocument.Parse(await response.Content.ReadAsStringAsync());
+
+            using var alipay_user_info_share_response = JsonDocument.Parse(payload.RootElement.GetProperty("alipay_user_info_share_response").ToString());
+
+            string status = alipay_user_info_share_response.RootElement.GetString("code");
+            if (status != "10000")
+            {
+                Logger.LogError("An error occurred while retrieving the user profile: the remote server " +
+                                "returned a {Status} response with the following message: {Message}.",
+                                /* Status: */ status,
+                                /* Message: */ payload.RootElement.GetString("msg"));
+
+                throw new HttpRequestException("An error occurred while retrieving user information.");
+            }
+
+            identity.AddClaim(new Claim(ClaimTypes.NameIdentifier, alipay_user_info_share_response.RootElement.GetString("user_id"), ClaimValueTypes.String, Options.ClaimsIssuer));
+
+            var principal = new ClaimsPrincipal(identity);
+            var context = new OAuthCreatingTicketContext(principal, properties, Context, Scheme, Options, Backchannel, tokens, alipay_user_info_share_response.RootElement);
+            context.RunClaimActions();
+
+            await Options.Events.CreatingTicket(context);
+            return new AuthenticationTicket(context.Principal, context.Properties, Scheme.Name);
+        }
+
+        protected override string FormatScope() => string.Join(",", Options.Scope);
+        private IDictionary<string, string> RSA2Sign(SortedDictionary<string, string> source)
+        {
+            var textparams = source.Where(p => !string.IsNullOrEmpty(p.Value)).Select(p => p.Key + "=" + p.Value);
+            if (!textparams.Any())
+                throw new HttpRequestException("sign fail");
+            var rsa2 = new RSAHelper(RSAType.RSA2, Encoding.UTF8, Options.ClientSecret, Options.AlipayPublicKey);
+            source.Add("sign", rsa2.Sign(string.Join("&", textparams)));
+            return source;
```

Comment:
This is confusing - at the call site it appears to create a new object, but it's actually mutating the state and returning the original. It would be clearer for this method to be called something like `SignParameters()` and return the value. Then the caller can append the sign parameter there.